File: 백준/Silver/1874. 스택 수열/스택 수열.cc
// stack

#include <iostream>
#include <stack>
#include <vector>

using namespace std;

int n;
int num = 1;
stack<int> st;
int arr[100001];
vector<char> ans;

int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    cin >> n;
    for (int i = 0; i < n; i++) cin >> arr[i];

    int idx = 0;
    while (num <= n) {
        while (arr[idx] >= num) {
            st.push(num++);
            ans.push_back('+');
        }
        while (!st.empty() && arr[idx] == st.top()) {
            st.pop();
            ans.push_back('-');
            idx++;
        }
        if (!st.empty() && arr[idx] < st.top()) break;
    }

    if (!st.empty()) {
        cout << "NO";
        return 0;
    }

    for (auto i : ans) {
        cout << i << "\n";
    }

    return 0;
}

File: 백준/Silver/1874. 스택 수열/스택 수열.py
# 1 2 3 4
# 1 2 3
# 1 2
# 1 2 5 6
# 1 2 5
# 1 2 5 7 8
# 1 2 5 7
# 1 2 
# 1
# -

# 현재 스택의 맨 위에 있는 수 >  뽑아야 하는 수 이면 불가능
# 스택 맨 위 = 뽑아야 하는 수 => pop 연산 (-)
# 스택 맨 위 < 뽑아야 하는 수 => 뽑아야 하는 수까지 push, 맨 위 pop

import sys

n = sys.stdin.readline()
st = []
ans = []
end = 0
idx = 1

for _ in range(int(n)):
    num = sys.stdin.readline()
    num = int(num)
    if len(st) == 0 or st[-1] != num:
        if idx > int(num):
            end = 1
            break
        while True:
            st.append(idx)
            ans.append("+")
            idx += 1
            if (st[-1] == num):
                st.pop()
                ans.append("-")
                break
    elif st[-1] == int(num):
        st.pop()
        ans.append("-")
        
if end == 1:
    print("NO")
else:
    for c in ans:
        print(c)